synapse/cryotank.py

```diff
@@ -571,3 +589,644 @@ def init(self, name, conf=None, timeout=None):
+        '''
+        Persists the index info to the database
+
+        Args:
+            progressonly (bool): if True, only persists the progress (i.e. more dynamic) information
+            txn (Optional[lmdb.Transaction]): if not None, will use that transaction to record data.  txn is
+            not committed.
+        Returns:
+            None
+        '''
+        d = {'delete': self.deleting,
+             'present': {k: metaentry.en() for k, metaentry in self.indices.items()}}
+
+        with contextlib.ExitStack() as stack:
+            if txn is None:
+                txn = stack.enter_context(self._dbenv.begin(db=self._metatbl, buffers=True, write=True))
+            if not progressonly:
+                txn.put(b'indices', s_msgpack.en(d), db=self._metatbl)
+            txn.put(b'progress', s_msgpack.en(self.progresses), db=self._metatbl)
+
+    def lowestProgress(self):
+        '''
+        Returns:
+            int: The next offset that should be indexed, based on active indices.
+        '''
+        nextoffsets = [p['nextoffset'] for iid, p in self.progresses.items() if not self.asleep[iid]]
+        return min(nextoffsets) if nextoffsets else s_lmdb.MAX_INT_VAL
+
+    def iidFromProp(self, prop):
+        '''
+        Args:
+            prop (str) The name of the indexed property
+        Returns:
+            int: the index id for the propname, None if not found
+        '''
+        return next((k for k, idx in self.indices.items() if idx.propname == prop), None)
+
+    def addIndex(self, prop, syntype, datapath, *args):
+        '''
+        Adds an index to the cryotank.
+
+        Args:
+            prop (str):  the name of the property this will be stored as in the normalized record
+            syntype (str):  the synapse type this will be interpreted as
+            datapath (str):  the datapath spec against which the raw record is run to extract a single field
+        that is passed to the type normalizer.
+            *args (str):  additional datapaths that will be tried in order if the first isn't present.
+        Returns:
+            None
+        Note:
+            Additional datapaths will be tried iff prior datapaths are not present, and *not* if
+            the normalization fails.
+        '''
+        if self.iidFromProp(prop) is not None:
+            raise ValueError('index already added')
+
+        s_datamodel.tlib.reqDataType(syntype)
+        iid = int.from_bytes(os.urandom(8), 'little')
+        self.indices[iid] = _MetaEntry(propname=prop, syntype=syntype, datapaths=(datapath,) + args)
+        self.progresses[iid] = {'nextoffset': 0, 'ngood': 0, 'nnormfail': 0}
+        self.persist()
+
+    def delIndex(self, prop):
+        '''
+        Deletes an index
+
+        Args:
+            prop (str): the (normalized) property name
+        Returns:
+            None
+        '''
+        iid = self.iidFromProp(prop)
+        if iid is None:
+            raise ValueError('Index not present')
```

Comment:
NoSuchIndx seems like a good idea :)

Reply:
Pretty much every ValueError should be replaced with some form of synerr so we can shovel in the missing prop in a `valu` field

Reply:
ok